functional_programming_tests/test/main.cpp

```diff
@@ -166,79 +183,117 @@ std_msgs::msg::UInt8MultiArray createUInt8MultiArrayMessageFromPath(
+ */
+std::map<error, std::string> const error_description = {
+    {error::NO_OCCUPANCY_MAP, "The Occupancy Map is empty."},
+    {error::INVALID_START_SIZE,
+     "The start field in the request is not of size 2."},
+    {error::INVALID_GOAL_SIZE,
+     "The goal field in the request is not of size 2."},
+    {error::NO_VALID_PATH,
+     "There is no valid path between the start and goal."}};
+
+/**
+ * @brief      Converts between types and dispatches to path generator
+ *
+ * @param[in]  request contains the start and goal positions
+ * @param[in]  occupancy_map to path through
+ * @param[in]  path_generator algorithm to use for pathing
+ *
+ * @return     The path if successful, otherwise an error
+ */
+tl::expected<GetPath::Response, error> callback(
+    std::shared_ptr<GetPath::Request> const request,
+    Map<unsigned char> const& occupancy_map,
+    pathing::PathingGenerator path_generator) {
   if (occupancy_map.get_data().size() == 0) {
-    return empty_response();
+    return tl::unexpected(error::NO_OCCUPANCY_MAP);
   }
   // Check to make sure start and goal fields of the request are of size 2
   if (request->start.data.size() != 2) {
-    return empty_response();
+    return tl::unexpected(error::INVALID_START_SIZE);
   }
   if (request->goal.data.size() != 2) {
-    return empty_response();
+    return tl::unexpected(error::INVALID_GOAL_SIZE);
   }
 
   auto const start = Position{request->start.data[0], request->start.data[1]};
   auto const goal = Position{request->goal.data[0], request->goal.data[1]};
 
   // Generate the path using the path generator function that was input
-  auto const path = path_generater(start, goal, occupancy_map);
-
+  auto const path = path_generator(start, goal, occupancy_map);
   if (!path.has_value()) {
-    return empty_response();
+    return tl::unexpected(error::NO_VALID_PATH);
   }
 
-  auto const response =
-      std::make_shared<example_srvs::srv::GetPath::Response>();
-  response->success.data = path.has_value();
-  response->path =
+  auto response = GetPath::Response{};
+  response.success.data = path.has_value();
+  response.path =
       pathing::utilities::createUInt8MultiArrayMessageFromPath(path.value());
 
   return response;
 }
 
-struct MainObject {
-  MainObject()
-      : set_costmap_wrapper_{[this](const std::shared_ptr<
-                                        example_srvs::srv::SetMap::Request>
-                                        request,
-                                    std::shared_ptr<
-                                        example_srvs::srv::SetMap::Response>
-                                        response) -> void {
-          const auto path = pathing::utilities::parseSetMapRequest(request);
+};  // namespace generate_path
+
+struct PathingManager {
+  /**
+   * @brief Manages occupancy_map and path generation for the pathing node
+   */
+  PathingManager()
+      : set_costmap_wrapper_{[this](auto const request, auto response) -> void {
```

Comment:
templated lambda via auto? I was getting type deduction errors at some point so I just replaced the autos with the concrete type.

Reply:
@griswaldbrooks

Reply:
it's able to deduce it because the type it's being assigned to is declared. this doesn't always work with ros stuff though. in this case it did :shrug: 
What I couldn't do was eliminate the return type and still have a nice `return` statement